tuf/signed/sign_test.go

```diff
@@ -264,9 +267,87 @@ func TestSignWithX509(t *testing.T) {
 	// RSAx509 key)
 	mockCryptoService := &StrictMockCryptoService{MockCryptoService{privKey}}
 	testData := data.Signed{}
-	err = Sign(mockCryptoService, &testData, tufRSAx509Key)
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		tufRSAx509Key,
+	)
+	err = Sign(mockCryptoService, &testData, baseRole)
 	assert.NoError(t, err)
 
 	assert.Len(t, testData.Signatures, 1)
 	assert.Equal(t, tufRSAx509Key.ID(), testData.Signatures[0].KeyID)
 }
+
+func TestSignWithOldKey(t *testing.T) {
+	cs := NewEd25519()
+	testData := data.Signed{}
+
+	key1, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+	)
+	Sign(cs, &testData, baseRole)
+
+	key2, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole = data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key2,
+	)
+	// should remove key1 sig even though it's valid. It no longer appears
+	// in the list of signing keys for the role
+	Sign(
+		cs,
+		&testData,
+		baseRole,
+	)
+
+	assert.Len(t, testData.Signatures, 1)
+	assert.Equal(t, key2.ID(), testData.Signatures[0].KeyID)
+}
+
+func TestSignWithOldSig(t *testing.T) {
+	cs := NewEd25519()
+	testData := data.Signed{}
+
+	key1, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	baseRole := data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+	)
+	Sign(cs, &testData, baseRole)
+
+	// we need cs to "forget" key1 so we can't sign with it
+	cs = NewEd25519()
+	key2, err := cs.Create(data.CanonicalRootRole, data.ED25519Key)
+	assert.NoError(t, err)
+
+	// modify test data to invalidate key1 sig
+	testData.Signed = []byte{0xff}
+
+	baseRole = data.NewBaseRole(
+		data.CanonicalRootRole,
+		1,
+		key1,
+		key2,
+	)
+	// should remove key1 sig because it's out of date
+	Sign(
+		cs,
+		&testData,
+		baseRole,
+	)
+
+	assert.Len(t, testData.Signatures, 1)
+	assert.Equal(t, key2.ID(), testData.Signatures[0].KeyID)
+}
```

Comment:
`Sign` is back to just taking `data.PublicKey...` The logic for `Sign` has always been meant to be all keys for a role are always passed, and we sign with whatever keys we have. I've seen that we haven't always obeyed that logic (I had to update one of the key rotation tests), and I don't see it documented anywhere obvious. I'll update the comment on `Sign`. This tests checks that it correctly removes a signature of a key that isn't passed and other tests in this file check it correctly preserves a (still valid) signature for a key that was passed but wasn't found in the private key store.